### PR TITLE
Mold: 0.9.2 -> 0.9.3, apply patch: "Intercept all invocations of ld, ld.gold or ld.lld"

### PR DESCRIPTION
**Pull request by jakobrs**
###### Motivation for this change
https://github.com/rui314/mold/issues/101, https://github.com/NixOS/nixpkgs/pull/128889#issuecomment-887736966

###### Things done

I've confirmed that `mold --run cargo build` uses mold as the linker.

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/tools/mold/default.nix [hunk 1]
```diff
@@ -1,5 +1,6 @@
 { stdenv
```
Comment (rui314) [inline pkgs/development/tools/mold/default.nix:1]: mold no longer relies on an external `tbb`. `tbb` is now included as a subdirectory of the project and statically-linked to mold executable.

Comment (jakobrs) [reply pkgs/development/tools/mold/default.nix:1]: Fixed, ty.

**Review by nitsky — APPROVED** [review]

**Review by nitsky — APPROVED** [review]

**Comment by jakobrs** [comment]
@SuperSandro2000 hope I'm not bothering you but since you were the merger of the other PR, could you merge this too?

**Comment by happysalada** [comment]
hey, just dropping to say that I've tried to update mold to the latest version to remove the patch (0.9.4) and it fails on darwin.
(the failure is related to endian.h). Checking the source, it seems removed on latest master.
I've tried building latest master and it fails with 
```
mold-unstable> In file included from third-party/tbb/include/tbb/../oneapi/tbb/task.h:22:
mold-unstable> third-party/tbb/include/tbb/../oneapi/tbb/detail/_task.h:221:13: error: aligned deallocation function of type 'void (void *, std::align_val_t) noexcept' is only available on macOS 10.14 or newer
mold-unstable>     virtual ~task() = default;
mold-unstable>             ^
mold-unstable> third-party/tbb/include/tbb/../oneapi/tbb/parallel_for.h:66:8: note: in defaulted destructor for 'tbb::detail::d1::task' first required here
mold-unstable> struct start_for : public task {
mold-unstable>        ^
mold-unstable> third-party/tbb/include/tbb/../oneapi/tbb/parallel_for.h:66:8: note: in implicit destructor for 'tbb::detail::d1::start_for<tbb::detail::d1::blocked_range<long long>, tbb::detail::d1::parallel_for_body_wrapper<(lambda at elf/compress.cc:72:49), long long>, const tbb::detail::d1::auto_partitioner>' first required here
mold-unstable> third-party/tbb/include/tbb/../oneapi/tbb/parallel_for.h:247:5: note: in instantiation of template class 'tbb::detail::d1::start_for<tbb::detail::d1::blocked_range<long long>, tbb::detail::d1::parallel_for_body_wrapper<(lambda at elf/compress.cc:72:49), long long>, const tbb::detail::d1::auto_partitioner>' requested here
mold-unstable>     start_for<Range,Body,const auto_partitioner>::run(range,body,partitioner);
mold-unstable>     ^
mold-unstable> third-party/tbb/include/tbb/../oneapi/tbb/parallel_for.h:316:9: note: in instantiation of function template specialization 'tbb::detail::d1::parallel_for<tbb::detail::d1::blocked_range<long long>, tbb::detail::d1::parallel_for_body_wrapper<(lambda at elf/compress.cc:72:49), long long>>' requested here
mold-unstable>         parallel_for(range, body, partitioner);
mold-unstable>         ^
mold-unstable> third-party/tbb/include/tbb/../oneapi/tbb/parallel_for.h:355:5: note: in instantiation of function template specialization 'tbb::detail::d1::parallel_for_impl<long long, (lambda at elf/compress.cc:72:49), const tbb::detail::d1::auto_partitioner>' requested here
mold-unstable>     parallel_for_impl<Index,Function,const auto_partitioner>(first, last, static_cast<Index>(1), f, auto_partitioner());
mold-unstable>     ^
mold-unstable> elf/compress.cc:72:8: note: in instantiation of function template specialization 'tbb::detail::d1::parallel_for<long long, (lambda at elf/compress.cc:72:49)>' requested here
mold-unstable>   tbb::parallel_for((i64)0, (i64)inputs.size(), [&](i64 i) {
mold-unstable>        ^
mold-unstable> third-party/tbb/include/tbb/../oneapi/tbb/detail/_task.h:221:13: note: if you supply your own aligned allocation functions, use -faligned-allocation to silence this diagnostic
mold-unstable>     virtual ~task() = default;
mold-unstable>             ^
mold-unstable> In file included from elf/compress.cc:17:
mold-unstable> In file included from third-party/tbb/include/tbb/parallel_for_each.h:17:
mold-unstable> In file included from third-party/tbb/include/tbb/../oneapi/tbb/parallel_for_each.h:27:
mold-unstable> third-party/tbb/include/tbb/../oneapi/tbb/parallel_for.h:66:8: error: aligned deallocation function of type 'void (void *, std::align_val_t) noexcept' is only available on macOS 10.14 or newer
mold-unstable> struct start_for : public task {
mold-unstable>        ^
mold-unstable> third-party/tbb/include/tbb/../oneapi/tbb/parallel_for.h:66:8: note: in implicit destructor for 'tbb::detail::d1::start_for<tbb::detail::d1::blocked_range<long long>, tbb::detail::d1::parallel_for_body_wrapper<(lambda at elf/compress.cc:72:49), long long>, const tbb::detail::d1::auto_partitioner>' first required here
mold-unstable> third-party/tbb/include/tbb/../oneapi/tbb/parallel_for.h:247:5: note: in instantiation of template class 'tbb::detail::d1::start_for<tbb::detail::d1::blocked_range<long long>, tbb::detail::d1::parallel_for_body_wrapper<(lambda at elf/compress.cc:72:49), long long>, const tbb::detail::d1::auto_partitioner>' requested here
mold-unstable>     start_for<Range,Body,const auto_partitioner>::run(range,body,partitioner);
mold-unstable>     ^
mold-unstable> third-party/tbb/include/tbb/../oneapi/tbb/parallel_for.h:316:9: note: in instantiation of function template specialization 'tbb::detail::d1::parallel_for<tbb::detail::d1::blocked_range<long long>, tbb::detail::d1::parallel_for_body_wrapper<(lambda at elf/compress.cc:72:49), long long>>' requested here
mold-unstable>         parallel_for(range, body, partitioner);
mold-unstable>         ^
mold-unstable> third-party/tbb/include/tbb/../oneapi/tbb/parallel_for.h:355:5: note: in instantiation of function template specialization 'tbb::detail::d1::parallel_for_impl<long long, (lambda at elf/compress.cc:72:49), const tbb::detail::d1::auto_partitioner>' requested here
mold-unstable>     parallel_for_impl<Index,Function,const auto_partitioner>(first, last, static_cast<Index>(1), f, auto_partitioner());
mold-unstable>     ^
mold-unstable> elf/compress.cc:72:8: note: in instantiation of function template specialization 'tbb::detail::d1::parallel_for<long long, (lambda at elf/compress.cc:72:49)>' requested here
mold-unstable>   tbb::parallel_for((i64)0, (i64)inputs.size(), [&](i64 i) {
mold-unstable>        ^
mold-unstable> third-party/tbb/include/tbb/../oneapi/tbb/parallel_for.h:66:8: note: if you supply your own aligned allocation functions, use -faligned-allocation to silence this diagnostic
mold-unstable> struct start_for : public task {
mold-unstable>        ^
mold-unstable> third-party/tbb/include/tbb/../oneapi/tbb/parallel_for.h:66:8: error: aligned deallocation function of type 'void (void *, std::align_val_t) noexcept' is only available on macOS 10.14 or newer
mold-unstable> struct start_for : public task {
mold-unstable>        ^
mold-unstable> third-party/tbb/include/tbb/../oneapi/tbb/parallel_for.h:66:8: note: in implicit destructor for 'tbb::detail::d1::start_for<tbb::detail::d1::blocked_range<long long>, tbb::detail::d1::parallel_for_body_wrapper<(lambda at elf/compress.cc:101:49), long long>, const tbb::detail::d1::auto_partitioner>' first required here
mold-unstable> third-party/tbb/include/tbb/../oneapi/tbb/parallel_for.h:247:5: note: in instantiation of template class 'tbb::detail::d1::start_for<tbb::detail::d1::blocked_range<long long>, tbb::detail::d1::parallel_for_body_wrapper<(lambda at elf/compress.cc:101:49), long long>, const tbb::detail::d1::auto_partitioner>' requested here
mold-unstable>     start_for<Range,Body,const auto_partitioner>::run(range,body,partitioner);
mold-unstable>     ^
mold-unstable> third-party/tbb/include/tbb/../oneapi/tbb/parallel_for.h:316:9: note: in instantiation of function template specialization 'tbb::detail::d1::parallel_for<tbb::detail::d1::blocked_range<long long>, tbb::detail::d1::parallel_for_body_wrapper<(lambda at elf/compress.cc:101:49), long long>>' requested here
mold-unstable>         parallel_for(range, body, partitioner);
mold-unstable>         ^
mold-unstable> third-party/tbb/include/tbb/../oneapi/tbb/parallel_for.h:355:5: note: in instantiation of function template specialization 'tbb::detail::d1::parallel_for_impl<long long, (lambda at elf/compress.cc:101:49), const tbb::detail::d1::auto_partitioner>' requested here
mold-unstable>     parallel_for_impl<Index,Function,const auto_partitioner>(first, last, static_cast<Index>(1), f, auto_partitioner());
mold-unstable>     ^
mold-unstable> elf/compress.cc:101:8: note: in instantiation of function template specialization 'tbb::detail::d1::parallel_for<long long, (lambda at elf/compress.cc:101:49)>' requested here
mold-unstable>   tbb::parallel_for((i64)0, (i64)shards.size(), [&](i64 i) {
mold-unstable>        ^
mold-unstable> third-party/tbb/include/tbb/../oneapi/tbb/parallel_for.h:66:8: note: if you supply your own aligned allocation functions, use -faligned-allocation to silence this diagnostic
mold-unstable> struct start_for : public task {
mold-unstable>        ^
mold-unstable> third-party/tbb/include/tbb/../oneapi/tbb/parallel_for.h:66:8: error: aligned deallocation function of type 'void (void *, std::align_val_t) noexcept' is only available on macOS 10.14 or newer
mold-unstable> struct start_for : public task {
mold-unstable>        ^
mold-unstable> third-party/tbb/include/tbb/../oneapi/tbb/parallel_for.h:66:8: note: in implicit destructor for 'tbb::detail::d1::start_for<tbb::detail::d1::blocked_range<long long>, tbb::detail::d1::parallel_for_body_wrapper<(lambda at elf/compress.cc:120:49), long long>, const tbb::detail::d1::auto_partitioner>' first required here
mold-unstable> third-party/tbb/include/tbb/../oneapi/tbb/parallel_for.h:247:5: note: in instantiation of template class 'tbb::detail::d1::start_for<tbb::detail::d1::blocked_range<long long>, tbb::detail::d1::parallel_for_body_wrapper<(lambda at elf/compress.cc:120:49), long long>, const tbb::detail::d1::auto_partitioner>' requested here
mold-unstable>     start_for<Range,Body,const auto_partitioner>::run(range,body,partitioner);
mold-unstable>     ^
mold-unstable> third-party/tbb/include/tbb/../oneapi/tbb/parallel_for.h:316:9: note: in instantiation of function template specialization 'tbb::detail::d1::parallel_for<tbb::detail::d1::blocked_range<long long>, tbb::detail::d1::parallel_for_body_wrapper<(lambda at elf/compress.cc:120:49), long long>>' requested here
mold-unstable>         parallel_for(range, body, partitioner);
mold-unstable>         ^
mold-unstable> third-party/tbb/include/tbb/../oneapi/tbb/parallel_for.h:355:5: note: in instantiation of function template specialization 'tbb::detail::d1::parallel_for_impl<long long, (lambda at elf/compress.cc:120:49), const tbb::detail::d1::auto_partitioner>' requested here
mold-unstable>     parallel_for_impl<Index,Function,const auto_partitioner>(first, last, static_cast<Index>(1), f, auto_partitioner());
mold-unstable>     ^
mold-unstable> elf/compress.cc:120:8: note: in instantiation of function template specialization 'tbb::detail::d1::parallel_for<long long, (lambda at elf/compress.cc:120:49)>' requested here
mold-unstable>   tbb::parallel_for((i64)0, (i64)inputs.size(), [&](i64 i) {
mold-unstable>        ^
mold-unstable> third-party/tbb/include/tbb/../oneapi/tbb/parallel_for.h:66:8: note: if you supply your own aligned allocation functions, use -faligned-allocation to silence this diagnostic
mold-unstable> struct start_for : public task {
mold-unstable>        ^
mold-unstable> third-party/tbb/include/tbb/../oneapi/tbb/parallel_for.h:66:8: error: aligned deallocation function of type 'void (void *, std::align_val_t) noexcept' is only available on macOS 10.14 or newer
mold-unstable> struct start_for : public task {
mold-unstable>        ^
mold-unstable> third-party/tbb/include/tbb/../oneapi/tbb/parallel_for.h:66:8: note: in implicit destructor for 'tbb::detail::d1::start_for<tbb::detail::d1::blocked_range<long long>, tbb::detail::d1::parallel_for_body_wrapper<(lambda at elf/compress.cc:154:49), long long>, const tbb::detail::d1::auto_partitioner>' first required here
mold-unstable> third-party/tbb/include/tbb/../oneapi/tbb/parallel_for.h:247:5: note: in instantiation of template class 'tbb::detail::d1::start_for<tbb::detail::d1::blocked_range<long long>, tbb::detail::d1::parallel_for_body_wrapper<(lambda at elf/compress.cc:154:49), long long>, const tbb::detail::d1::auto_partitioner>' requested here
mold-unstable>     start_for<Range,Body,const auto_partitioner>::run(range,body,partitioner);
mold-unstable>     ^
mold-unstable> third-party/tbb/include/tbb/../oneapi/tbb/parallel_for.h:316:9: note: in instantiation of function template specialization 'tbb::detail::d1::parallel_for<tbb::detail::d1::blocked_range<long long>, tbb::detail::d1::parallel_for_body_wrapper<(lambda at elf/compress.cc:154:49), long long>>' requested here
mold-unstable>         parallel_for(range, body, partitioner);
mold-unstable>         ^
mold-unstable> third-party/tbb/include/tbb/../oneapi/tbb/parallel_for.h:355:5: note: in instantiation of function template specialization 'tbb::detail::d1::parallel_for_impl<long long, (lambda at elf/compress.cc:154:49), const tbb::detail::d1::auto_partitioner>' requested here
mold-unstable>     parallel_for_impl<Index,Function,const auto_partitioner>(first, last, static_cast<Index>(1), f, auto_partitioner());
mold-unstable>     ^
mold-unstable> elf/compress.cc:154:8: note: in instantiation of function template specialization 'tbb::detail::d1::parallel_for<long long, (lambda at elf/compress.cc:154:49)>' requested here
mold-unstable>   tbb::parallel_for((i64)0, (i64)shards.size(), [&](i64 i) {
mold-unstable>        ^
mold-unstable> third-party/tbb/include/tbb/../oneapi/tbb/parallel_for.h:66:8: note: if you supply your own aligned allocation functions, use -faligned-allocation to silence this diagnostic
mold-unstable> struct start_for : public task {
mold-unstable>        ^
mold-unstable> 5 errors generated.
mold-unstable> make: *** [<builtin>: elf/compress.o] Error 1
error: builder for '/nix/store/n11f4jqc10d1w51xdpzpwi842fvx81rw-mold-unstable-2021-09-05.drv' failed with exit code 2;
       last 10 log lines:
       >     parallel_for_impl<Index,Function,const auto_partitioner>(first, last, static_cast<Index>(1), f, auto_partitioner());
       >     ^
       > elf/compress.cc:154:8: note: in instantiation of function template specialization 'tbb::detail::d1::parallel_for<long long, (lambda at elf/compress.cc:154:49)>' requested here
       >   tbb::parallel_for((i64)0, (i64)shards.size(), [&](i64 i) {
       >        ^
       > third-party/tbb/include/tbb/../oneapi/tbb/parallel_for.h:66:8: note: if you supply your own aligned allocation functions, use -faligned-allocation to silence this diagnostic
       > struct start_for : public task {
       >        ^
       > 5 errors generated.
       > make: *** [<builtin>: elf/compress.o] Error 1
       For full logs, run 'nix log /nix/store/n11f4jqc10d1w51xdpzpwi842fvx81rw-mold-unstable-2021-09-05.drv'.
```

it looks like an update would have to mark it broken on darwin, that's why I didn't make a PR.

just saying that I think the project is interesting, feel free to ping me if you need reviews for PRs to update it.

**Comment by rui314** [comment]
Note that mold 0.9.4 can't be built on Darwin per se, and even if you can, it can't be used to create macOS binaries. It can only create ELF (Linux) binaries. So you can use it for cross compilation but not for macOS native development.

**Comment by happysalada** [comment]
Got it, thanks for chiming in. In the next PR the platforms attribute should be adjusted accordingly then.

**Comment by 06kellyjac** [comment]
Anyone tried 0.9.6?

**Comment by jakobrs** [comment]
Version update: #139920

**Pull request merged**